modules/Evse15118D20/charger/ISO15118_chargerImpl.cpp

```diff
@@ -474,6 +584,56 @@ iso15118::session::feedback::Callbacks ISO15118_chargerImpl::create_callbacks()
 
     callbacks.selected_protocol = [this](const std::string& protocol) { publish_selected_protocol(protocol); };
 
+    callbacks.selected_vas_services = [this](const dt::VasSelectedServiceList& selected_services) {
+        // Group selected services by VAS provider index
+        std::map<size_t, std::vector<types::iso15118_vas::SelectedService>> services_by_provider;
+
+        for (const auto& service : selected_services) {
+            const auto provider_index = get_vas_provider_index(service.service_id);
+            if (not provider_index.has_value()) {
+                EVLOG_warning << "Selected Service ID " << service.service_id
+                              << " is not supported by any VAS provider";
+                continue;
+            }
+
+            types::iso15118_vas::SelectedService converted_service;
+            converted_service.service_id = service.service_id;
+            converted_service.parameter_set_id = service.parameter_set_id;
+
+            services_by_provider[*provider_index].push_back(converted_service);
```

Comment:
Why not call `mod->r_iso15118_vas[provider_index]->call_selected_services(converted_service);` directly here?

Reply:
Sry I dont get it. There is no single call with all selected services. For every selected service there is an extra call to the provider.

Reply:
Right now there is one small difference:
Your approach:
```
provider[0]->call(34)
provider[0]->call(35)
provider[1]->call(6000)
```
My suggestion (Depends on selected_services):
```
provider[0]->call(34)
provider[1]->call(6000)
provider[0]->call(35)
```
I don't think that's a problem for the VAS modules.

Reply:
The selected services are first grouped by provider index, this is the code above. It is stored in the map `services_by_provider`. Your example would result in the map `{0: [34, 35], 1: [6000]}`. After grouping, the providers are notified in the code below, resulting in the two calls `provider[0]->call([34,35]); provider[1]->call([6000])`. The call itself is happening in the code below, lines 604-606